### PR TITLE
Add edge lengths

### PR DESCRIPTION
**Pull request by streeve**
(From GCNN:)
With @pzhanggit 's help I have found:
- Only PNA and CGConv support
- GIN has an additional version GINE for edges
~~PNA currently breaks because we don't define the `edge_encoder`~~
~~GINE currently breaks because dimensions that don't agree (but I think actually do based on what I've checked)~~

Only implemented for PNA and CG (GINE does not have robust enough support to justify adding). New test only calls for these two models with edge lengths

Part of #2

### CODE REVIEW

**Comment by streeve** [comment]
Waiting for review on this one

**Comment by allaffa** [comment]
@streeve 
Reviewing this will be my task on Jira this week :)

File: hydragnn/unit_tests/test_graphs.py [hunk 1]
```diff
@@ -33,6 +31,10 @@ def pytest_train_model(model_type, ci_input, overwrite_data=False):
         config = json.load(f)
     config["NeuralNetwork"]["Architecture"]["model_type"] = model_type
 
+    # Only run with edge lengths for models that support them.
+    if use_lengths:
+        config["NeuralNetwork"]["Architecture"]["edge_features"] = ["lengths"]
```
Comment (streeve) [inline hydragnn/unit_tests/test_graphs.py:36]: @allaffa I add the edge feature on the fly here

File: hydragnn/unit_tests/test_graphs.py [hunk 2]
```diff
@@ -73,7 +75,7 @@ def pytest_train_model(model_type, ci_input, overwrite_data=False):
 
     # Since the config file uses PNA already, test the file overload here.
     # All the other models need to use the locally modified dictionary.
-    if model_type == "PNA":
+    if model_type == "PNA" and not use_lengths:
```
Comment (pzhanggit) [inline hydragnn/unit_tests/test_graphs.py:78]: The `if ` condition here is not needed.

Comment (streeve) [reply hydragnn/unit_tests/test_graphs.py:78]: But if we modify the config we can't use the original file

Comment (pzhanggit) [reply hydragnn/unit_tests/test_graphs.py:78]: We can use the config dictionary, right?

File: hydragnn/preprocess/serialized_dataset_loader.py [hunk 3]
```diff
@@ -66,15 +66,22 @@ def load_serialized_data(self, dataset_path: str, config):
             _ = pickle.load(f)
             dataset = pickle.load(f)
 
-        edge_index, edge_distances = self.__compute_edges(
+        edge_index, edge_length = self.__compute_edges(
```
Comment (pzhanggit) [inline hydragnn/preprocess/serialized_dataset_loader.py:69]: @allaffa  I think you are right about something wrong with the normalization. 
@streeve  Could you check line 196 inside self.__compute_edges(...)
`   # Normalize the lengths using min-max normalization
        edge_lengths = (edge_lengths - min(edge_lengths)) / (
            max(edge_lengths) - min(edge_lengths)
        )`. Currently, it uses min/max edge_length in each graph/data sample for normalization, which varies in dataset.

File: hydragnn/unit_tests/test_graphs.py (hunk 2 above)
Comment (streeve) [reply hydragnn/unit_tests/test_graphs.py:78]: Then we won't test the file interface and miss some bug

Comment (pzhanggit) [reply hydragnn/unit_tests/test_graphs.py:78]: Aha, I see what you mean. It is for testing the file overload. Thanks

File: hydragnn/preprocess/serialized_dataset_loader.py (hunk 3 above)
Comment (allaffa) [reply hydragnn/preprocess/serialized_dataset_loader.py:69]: Yep. the normalization should use `min` and `max` that span the entire dataset.

Comment (pzhanggit) [reply hydragnn/preprocess/serialized_dataset_loader.py:69]: Yeah, it is probably okay for now, given all the samples in the dataset have the same node position, but need to be improved.

**Comment by allaffa** [comment]
@streeve 
The configuration file `examples/configuraton.json` still causes an error when PNA is used. 
PNA currently breaks because we don't define the edge_encoder. 
I saw that this error was crossed out, so I assumed the issue was solved. But I still get this error when I run the version of the code on this branch.

**Comment by allaffa** [comment]
@streeve 

The constructor of the `PNA` model that overwrites the constructor of the `Base` model in `PNAStack.py` does not have edge attributes as input parameters. I suspect that the edge length is never used. Am I wrong?

File: hydragnn/preprocess/serialized_dataset_loader.py [hunk 4]
```diff
@@ -84,6 +97,15 @@ def load_serialized_data(self, dataset_path: str, config):
...
+        # Normalize the lengths using min-max normalization
+        for data in dataset:
+            if (
+                "edge_features" in config["Architecture"]
+                and config["Architecture"]["edge_features"]
+            ):
+                #data.edge_attr = (data.edge_attr - self.min_edge_lengh_global)/( self.max_edge_lengh_global - self.min_edge_lengh_global )
+                data.edge_attr = (data.edge_attr)/( self.max_edge_lengh_global )
+
         if "subsample_percentage" in config["Variables_of_interest"].keys():
```
Comment (allaffa) [inline hydragnn/preprocess/serialized_dataset_loader.py:109]: @streeve 
After PR#30 has been merged, I will rebase this one and use `torch_geometric.tramsforms.distance` to compute the edge length. This transformation also provide a Boolean variable to automatically normalize all the edge lengths ion the range (0,1) if it is set to True. 
https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/transforms/distance.html#Distance

File: hydragnn/models/Base.py [hunk 5]
```diff
@@ -194,7 +195,16 @@ def forward(self, data):
         )
         ### encoder part ####
         for conv, batch_norm in zip(self.convs, self.batch_norms):
-            x = F.relu(batch_norm(conv(x=x, edge_index=edge_index)))
+            if (data.edge_attr is not None) and (
```
Comment (pzhanggit) [inline hydragnn/models/Base.py:198]: Maybe move this outside of the for loop? We only need to assess data.edge_attr once.

File: hydragnn/models/Base.py [hunk 6]
```diff
@@ -195,7 +195,11 @@ def forward(self, data):
         )
         ### encoder part ####
         for conv, batch_norm in zip(self.convs, self.batch_norms):
-            if (data.edge_attr is not None) and (self.edge_dim is not None):
+            if (data.edge_attr is not None) and (
+                hasattr(self, "edge_dim")
```
Comment (streeve) [inline hydragnn/models/Base.py:199]: @allaffa is this for the edge case where the data has `edge_attr`, but the model doesn't support it?

File: hydragnn/models/Base.py (hunk 5 above)
Comment (streeve) [reply hydragnn/models/Base.py]: Done in the last commit

**Comment by streeve** [comment]
Cleaned up and rebased on main - thanks for the help @allaffa. This is ready for another review

**Comment by allaffa** [comment]
> Cleaned up and rebased on main - thanks for the help @allaffa. This is ready for another review @pzhanggit

Your commit history is so clean and every commit is so clearly mapped into changes made to the code...

File: hydragnn/models/Base.py (hunk 6 above)
Comment (allaffa) [reply hydragnn/models/Base.py]: Correct.

File: hydragnn/preprocess/serialized_dataset_loader.py [hunk 7]
```diff
@@ -67,9 +67,11 @@ def load_serialized_data(self, dataset_path: str, config):
             loop=False,
             max_num_neighbors=config["Architecture"]["max_neighbours"],
         )
+        compute_edge_lengths = Distance(norm=True, cat=True)
```
Comment (pzhanggit) [inline hydragnn/preprocess/serialized_dataset_loader.py:70]: Setting `norm=True` would normalize the distance with the maximum distance in each data sample. Is this by design?

Comment (allaffa) [reply hydragnn/preprocess/serialized_dataset_loader.py:70]: I was thinking Both this right now. I want to set it to `False`

File: hydragnn/models/Base.py [hunk 8]
```diff
@@ -192,9 +202,18 @@ def forward(self, data):
...
             data.batch,
         )
+        use_edge_attr = False
+        if (data.edge_attr is not None) and (self.use_edge_attr):
+            use_edge_attr = True
+
         ### encoder part ####
         for conv, batch_norm in zip(self.convs, self.batch_norms):
-            x = F.relu(batch_norm(conv(x=x, edge_index=edge_index)))
+            if use_edge_attr:
```
Comment (pzhanggit) [inline hydragnn/models/Base.py:211]: I was thinking a structure like,
```
if use_edge_attr:
       for conv, batch_norm in zip(self.convs, self.batch_norms):
             c = conv(x=x, edge_index=edge_index, edge_attr=data.edge_attr)
             x = F.relu(batch_norm(c))
else:
       for conv, batch_norm in zip(self.convs, self.batch_norms):
             c = conv(x=x, edge_index=edge_index)
             x = F.relu(batch_norm(c))
```
Not sure how much difference it could make though.

Comment (allaffa) [reply hydragnn/models/Base.py:211]: The last commit from Sam does that. The relu call is out of the if-else condition though. 

if use_edge_attr:
       for conv, batch_norm in zip(self.convs, self.batch_norms):
             c = conv(x=x, edge_index=edge_index, edge_attr=data.edge_attr)
else:
       for conv, batch_norm in zip(self.convs, self.batch_norms):
             c = conv(x=x, edge_index=edge_index)

 x = F.relu(batch_norm(c))

Comment (pzhanggit) [reply hydragnn/models/Base.py:211]: > 

Not really. For the current code, it still needs to run if-else for every `conv` inside for loop. [Btw, `relu` should be inside for loop in your comment. ]

Comment (allaffa) [reply hydragnn/models/Base.py]: @pzhanggit 
committed new changes to follow your suggestions. Take a look and let me know. I am debating now whether to NOT perform the normalization on edge_length at all or if we should perform it over the entire dataset

**Review by pzhanggit — CHANGES_REQUESTED** [review]
It looks good to me. Could you run one case with edge feature on DFT dataset? Wondering how much accuracy improvement it can get.

File: hydragnn/preprocess/serialized_dataset_loader.py (hunk 7 above)
Comment (pzhanggit) [reply hydragnn/preprocess/serialized_dataset_loader.py]: We probably still need normalization (otherwise edge feature likely will dominate everything,  given that all other features are normalized). Need to figure out a proper way to do the normalization.

File: hydragnn/models/Base.py (hunk 8 above)
Comment (pzhanggit) [reply hydragnn/models/Base.py]: Lol, see my comment on normalization above. In short, I think we should have normalization but need do it properly.

File: hydragnn/models/Base.py [hunk 9]
```diff
@@ -15,6 +15,7 @@
 from torch_geometric.nn import global_mean_pool, BatchNorm
 from torch.nn import GaussianNLLLoss
 import sys
+import inspect
```
Comment (streeve) [inline hydragnn/models/Base.py:18]: Remove?

File: hydragnn/preprocess/serialized_dataset_loader.py (hunk 7 above)
Comment (allaffa) [reply hydragnn/preprocess/serialized_dataset_loader.py]: Done.

File: hydragnn/preprocess/serialized_dataset_loader.py [hunk 10]
```diff
@@ -69,9 +69,21 @@ def load_serialized_data(self, dataset_path: str, config):
...
         compute_edge_lengths = Distance(norm=False, cat=True)
 
+        dataset[:]=[compute_edges(data) for data in dataset]
+        dataset[:]=[compute_edge_lengths(data) for data in dataset]
+
+        min_edge_length = torch.Tensor([float('inf')])
+        max_edge_length = torch.Tensor([float('-inf')])
+
+        for data in dataset:
+            min_edge_length = torch.min(min_edge_length, torch.min(data.edge_attr))
```
Comment (streeve) [inline hydragnn/preprocess/serialized_dataset_loader.py:79]: Comment out since we're not using?

File: hydragnn/preprocess/serialized_dataset_loader.py [hunk 11]
```diff
@@ -69,9 +69,21 @@ def load_serialized_data(self, dataset_path: str, config):
         )
         compute_edge_lengths = Distance(norm=False, cat=True)
 
+        dataset[:]=[compute_edges(data) for data in dataset]
```
Comment (streeve) [inline hydragnn/preprocess/serialized_dataset_loader.py]: This is nice

Comment (allaffa) [reply hydragnn/preprocess/serialized_dataset_loader.py]: @streeve Done.

File: hydragnn/preprocess/serialized_dataset_loader.py (hunk 10 above)
Comment (allaffa) [reply hydragnn/preprocess/serialized_dataset_loader.py]: @streeve Done.

**Comment by allaffa** [comment]
> It looks good to me. Could you run one case with edge feature on DFT dataset? Wondering how much accuracy improvement it can get.

@pzhanggit no improvement seen. I will open a separate PR to inver the edge length as edge attributes. Will keep you posted.

File: hydragnn/preprocess/serialized_dataset_loader.py [hunk 12]
```diff
@@ -67,9 +67,22 @@ def load_serialized_data(self, dataset_path: str, config):
             loop=False,
             max_num_neighbors=config["Architecture"]["max_neighbours"],
         )
+        compute_edge_lengths = Distance(norm=False, cat=True)
+
+        dataset[:] = [compute_edges(data) for data in dataset]
+        dataset[:] = [compute_edge_lengths(data) for data in dataset]
+
+        min_edge_length = torch.Tensor([float("inf")])
```
Comment (pzhanggit) [inline hydragnn/preprocess/serialized_dataset_loader.py:75]: Is it used anywhere?

Comment (streeve) [reply hydragnn/preprocess/serialized_dataset_loader.py]: Thanks, fixed

**Review by pzhanggit — APPROVED** [review]
